File: function_ir.py
from span import SourceSpan
from annotation_ir import AnnotationIR


class ParamIR:
    def __init__(self, symbol_id: int, name: str, annotation: AnnotationIR, default, span: SourceSpan):
        self.symbol_id = symbol_id
        self.name = name
        self.annotation = annotation
        self.default = default
        self.span = span

class FunctionIR:
    def __init__(
        self,
        id: int,
        symbol_id: int,
        name: str,
        scope_id: int,  # parent scope where function name is bound
        body_scope_id: int,  # function-local scope
        params: list[ParamIR],
        returns,
        decorators,
        span: SourceSpan,
    ):
        self.id = id
        self.symbol_id = symbol_id
        self.name = name
        self.scope_id = scope_id
        self.body_scope_id = body_scope_id
        self.params = params
        self.returns = returns
        self.decorators = decorators
        self.span = span

File: rust/src/ir/nodes/function_ir.rs
use crate::ir::{AnnotationIR, ExprIR, SourceSpan, StmtIR};

#[derive(Debug, Clone)]
pub struct ParamIR {
    pub symbol_id: i64,
    pub name: String,
    pub annotation: Option<AnnotationIR>,
    pub default: Option<Box<ExprIR>>,
    pub span: Option<SourceSpan>,
}

impl ParamIR {
    pub fn new(
        symbol_id: i64,
        name: impl Into<String>,
        annotation: Option<AnnotationIR>,
        default: Option<ExprIR>,
        span: Option<SourceSpan>,
    ) -> Self {
        Self {
            symbol_id,
            name: name.into(),
            annotation,
            default: default.map(Box::new),
            span,
        }
    }
}

#[derive(Debug, Clone)]
pub struct ReturnIR {
    pub value: Option<Box<ExprIR>>,
    pub span: Option<SourceSpan>,
}

impl ReturnIR {
    pub fn new(value: Option<ExprIR>, span: Option<SourceSpan>) -> Self {
        Self {
            value: value.map(Box::new),
            span,
        }
    }
}

#[derive(Debug, Clone)]
pub struct FunctionIR {
    pub id: i64,
    pub symbol_id: i64,
    pub name: String,
    pub scope_id: i64,
    pub body_scope_id: i64,
    pub params: Vec<ParamIR>,
    pub body: Vec<StmtIR>,
    pub returns: Option<AnnotationIR>,
    pub decorators: Vec<ExprIR>,
    pub span: Option<SourceSpan>,
}
